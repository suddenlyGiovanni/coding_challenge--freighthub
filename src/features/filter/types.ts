export default {
  SET_FILTER_SHIPMENTS: '@filter/SET_FILTER_SHIPMENTS',
} as const
