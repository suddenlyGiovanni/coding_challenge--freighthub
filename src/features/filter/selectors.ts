import { RootState } from 'typesafe-actions'

export const getFilter = (state: RootState) => state.filter
