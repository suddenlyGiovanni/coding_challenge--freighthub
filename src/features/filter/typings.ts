export interface Filter {
  name: string
  mode: {
    sea: boolean
    air: boolean
    rail: boolean
  }
  status: 'ACTIVE' | 'COMPLETED' | 'ALL'
}
