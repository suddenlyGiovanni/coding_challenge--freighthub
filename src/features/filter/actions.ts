import { createAction } from 'typesafe-actions'

import filterTypes from './types'
import { Filter } from './typings'

export const setFilter = createAction(
  filterTypes.SET_FILTER_SHIPMENTS,
  action => {
    return (filter: Partial<Filter>) => action(filter)
  }
)
