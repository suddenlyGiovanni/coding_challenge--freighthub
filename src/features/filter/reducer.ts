import { Reducer } from 'redux'
import { getType, ActionType } from 'typesafe-actions'

import * as filterActions from './actions'
import * as FilterTypes from './typings'

export type FilterAction = ActionType<typeof filterActions>

export type FilterState = Readonly<FilterTypes.Filter>

const initialState: FilterState = {
  name: '',
  mode: {
    air: true,
    sea: true,
    rail: true,
  },
  status: 'ALL',
}

export const filterReducer: Reducer<FilterState, FilterAction> = (
  state = initialState,
  action
) => {
  if (action.type === getType(filterActions.setFilter)) {
    return { ...state, ...action.payload }
  }
  return state
}
