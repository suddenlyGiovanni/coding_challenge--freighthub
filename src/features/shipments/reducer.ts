import { Reducer } from 'redux'
import { getType, ActionType } from 'typesafe-actions'

import * as shipmentsActions from './actions'
import * as ShipmentsTypes from './typings'

export type ShipmentsAction = ActionType<typeof shipmentsActions>

export type ShipmentsState = Readonly<ShipmentsTypes.Shipments>

const initialState: ShipmentsState = []

export const shipmentsReducer: Reducer<ShipmentsState, ShipmentsAction> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    case getType(shipmentsActions.setShipments): {
      return [...action.payload.shipments]
    }

    case getType(shipmentsActions.editShipmentName): {
      const { id, name } = action.payload
      return state.map(storeShipment => {
        return storeShipment.id === id
          ? { ...storeShipment, name } //
          : storeShipment //
      })
    }

    case getType(shipmentsActions.setShipment): {
      const { shipment } = action.payload
      return state.map(storeShipment => {
        return storeShipment.id === shipment.id
          ? shipment //
          : storeShipment //
      })
    }

    default:
      return state
  }
}
