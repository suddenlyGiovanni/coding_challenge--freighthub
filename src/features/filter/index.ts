import * as filterActions from './actions'
import { filterReducer } from './reducer'
import * as filterSelectors from './selectors'
import filterTypes from './types'

export { filterTypes, filterActions, filterSelectors, filterReducer }
