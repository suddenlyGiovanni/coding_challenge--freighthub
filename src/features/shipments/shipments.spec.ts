import {
  shipmentsActions as actions,
  shipmentsTypes as types,
  shipmentsReducer as reducer,
  shipmentsSelectors as selectors,
} from './index'

import { shipments, shipment } from 'test/mockDb'
import { ShipmentsState } from './reducer'
import { RootState } from 'typesafe-actions'

import { initialState as initialStateFilter } from 'features/filter/reducer'

/**
 * FIXTURES
 */
function getInitialState(initial?: ShipmentsState) {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return reducer(initial, {} as any)
}

function getRootState(initial?: ShipmentsState) {
  return { shipments: getInitialState(initial) }
}

describe('feature: `shipments`', () => {
  // ACTIONS CREATOR
  describe('actions', () => {
    // FETCH_SHIPMENTS
    it('`fetchShipments` should create an action with the correct signature', () => {
      expect(actions.fetchShipments()).toMatchObject({
        type: types.FETCH_SHIPMENTS,
      })
    })

    // SET_SHIPMENTS
    it('`setShipments` should create an action with the correct signature', () => {
      expect(actions.setShipments({ data: shipments })).toMatchObject({
        type: types.SET_SHIPMENTS,
        payload: { shipments },
      })
    })

    // EDIT_NAME
    it('`editShipmentName` should create an action with the correct signature', () => {
      const name = 'T-shirts from Shanghai to Hamburg'
      const id = 'U1002'
      expect(actions.editShipmentName({ name, id })).toMatchObject({
        type: types.EDIT_NAME,
        payload: { name, id },
      })
    })

    // SET_SHIPMENT
    it('`setShipment` should create an action with the correct signature', () => {
      expect(actions.setShipment({ data: shipment })).toMatchObject({
        type: types.SET_SHIPMENT,
        payload: { shipment },
      })
    })
  })

  // REDUCER
  describe('reducer', () => {
    it('initial state should match a snapshot', () => {
      const initialState = getInitialState()
      expect(initialState).toMatchSnapshot()
    })

    it('save to the store the list of all the shipments', () => {
      // Arrange
      const initialState = getInitialState()
      const newShipments = [...shipments, shipment]
      // Act
      const stateSnapshot1 = reducer(
        initialState,
        actions.setShipments({ data: shipments })
      )

      const stateSnapshot2 = reducer(
        initialState,
        actions.setShipments({ data: newShipments })
      )

      // Assert
      expect(stateSnapshot1).toHaveLength(shipments.length)
      expect(stateSnapshot1).toEqual(shipments)
      expect(stateSnapshot2).toEqual(newShipments)
    })

    it('updates the name field of a specific shipment', () => {
      // Arrange
      const initialState = getInitialState(shipments)
      const shipmentId = shipments[0].id
      const shipmentNewName = 'Really old spring collection'

      // Act
      const stateSnapshot1 = reducer(
        initialState,
        actions.editShipmentName({
          id: shipmentId,
          name: shipmentNewName,
        })
      )
      const editedEntry = stateSnapshot1.find(s => s.id === shipmentId)

      // Assert
      expect(stateSnapshot1).not.toEqual(initialState)
      expect((editedEntry && editedEntry.name) || '').toBe(shipmentNewName)
    })

    it('update a single shipment', () => {
      // Arrange
      const initialState = getInitialState(shipments)
      const newShipmentData = { ...initialState[0] }
      const shipmentNewName = 'Really old spring collection'
      newShipmentData.name = shipmentNewName

      // Act
      const stateSnapshot = reducer(
        initialState,
        actions.setShipment({ data: newShipmentData })
      )
      const editedEntry = stateSnapshot.find(s => s.id === newShipmentData.id)

      // Assert
      expect(stateSnapshot).not.toEqual(initialState)
      expect((editedEntry && editedEntry.name) || '').toBe(shipmentNewName)
    })

    // SELECTORS
    describe('selectors', () => {
      const initialRootState = getRootState(shipments)

      it('getShipments', () => {
        expect(selectors.getShipments(initialRootState as RootState)).toEqual(
          shipments
        )
      })

      it('getShipmentByID', () => {
        expect(
          selectors.getShipmentByID(
            initialRootState as RootState,
            shipments[0].id
          )
        ).toEqual(shipments[0])
      })

      it('getFilteredShipments', () => {
        const rootState = {
          ...initialRootState,
          filter: { ...initialStateFilter, status: 'COMPLETED' },
        }

        expect(
          selectors.getFilteredShipments(rootState as RootState)[0][0]
        ).toEqual(shipments[1])
      })
    })
  })
})
