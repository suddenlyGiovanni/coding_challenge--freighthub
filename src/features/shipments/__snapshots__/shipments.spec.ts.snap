// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feature: \`shipments\` reducer initial state should match a snapshot 1`] = `Array []`;
